dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -401,21 +405,26 @@ public UfsUrl join(String suffix) {
     while (nonEmptyIndex < suffixArray.length && suffixArray[nonEmptyIndex].isEmpty())  {
       nonEmptyIndex++;
     }
-    List<String> suffixComponentsList;
+    List<String> suffixComponentsList = Arrays.asList(suffixArray);
+    List<String> noEmptyElemSuffixComponentsList;
     if (nonEmptyIndex == 0) {
-      suffixComponentsList = Arrays.asList(suffixArray);
+      noEmptyElemSuffixComponentsList = suffixComponentsList;
     } else {
-      // TODO(Tony Sun): optimize the performance later
-      suffixComponentsList = Arrays.asList(
-          Arrays.copyOfRange(
-              suffixArray,
-              nonEmptyIndex, suffixArray.length));
+      noEmptyElemSuffixComponentsList = suffixComponentsList.subList(nonEmptyIndex,
+          suffixComponentsList.size());
     }
     List<String> pathComponents = mProto.getPathComponentsList();
     return new UfsUrl(UfsUrlMessage.newBuilder()
         .setScheme(mProto.getScheme())
         .setAuthority(mProto.getAuthority())
         .addAllPathComponents(pathComponents)
-        .addAllPathComponents(suffixComponentsList).build());
+        .addAllPathComponents(noEmptyElemSuffixComponentsList).build());
+  }
+
+  public UfsUrl joinUnsafe(String suffix) {
```

Comment:
add a precondition check to ensure `suffix` does not itself contain `/`.

Reply:
AlluxioURI does not check the suffix, maybe we should follow the design.

Reply:
We don't need to follow suit of AlluxioURI. We are literally rewriting it because of its bad design decisions.

1. A precondition check adds almost no extra performance cost.
2. I don't think simply giving it a name suffixed by `unsafe` justifies sacrificing correctness for performance, especially when it's a public method.
3. If this method has not concrete use cases now, I'd prefer it not included.